Bugfix/771 test error beforetest

sksamuel opened the pull request.

sksamuel commented on kotlintest-runner/kotlintest-runner-junit5/src/main/kotlin/io/kotlintest/runner/junit5/JUnitTestRunnerListener.kt:
```diff
@@ -193,9 +187,11 @@ class JUnitTestRunnerListener(private val listener: EngineExecutionListener,
-      val result = when (t) {
-        null -> TestExecutionResult.successful()
-        else -> TestExecutionResult.failed(t)
+      val firstTestError = results.find { it.result.error != null }?.result?.error
```
I am pretty sure that the JUnit specs say that a parent must pass even if it's children fail.
This is why I do not search the children for a failed test.

LeoColman replied:
Perhaps. Let me see what I can do.

LeoColman replied:
The thing is, if we don't fail the parent, the test case won't ever be created

LeoColman replied:
and it won't show in IntelliJ, that's why this fix was necessary

LeoColman replied:
perhaps we can add the dummy failure like we do with our runner

sksamuel commented on kotlintest-runner/kotlintest-runner-console/src/main/kotlin/io/kotlintest/runner/console/TeamCityConsoleWriter.kt:
```diff
@@ -1,17 +1,14 @@
 
+  private val results = HashSet<TestResult>()
+
   private var errors = false
```
Is this read anywhere?

LeoColman committed:
Fixes tests passing when before test fails

The problem with this issue was that any exception that happened in either `BeforeTest` or `AfterTest` was being silently ignored due to Coroutine Launch instead of `async`.

This commit fixes that error, and also modifies listener execution to notify `JUnit` and our executor to be able to identify these errors and show that in the interface.

Part of #771
